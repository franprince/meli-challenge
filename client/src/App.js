import "./App.scss";
import React, { useState } from "react";
import Product from "./components/Product";
import Header from "./components/Header";
import SearchBox from "./components/SearchBox";

function App() {
  const [productData, setProductData] = useState({
    data: null,
    status: "idle",
    error: "",
  });

  async function handleSearch(event, value) {
    event.preventDefault();
    try {
      if (!value) return;
      setProductData({ ...productData, status: "pending" });
      const res = await fetch(`http://localhost:3001/products/${value}`);
      const data = await res.json();
      if (data.error) {
        setProductData({
          data: null,
          status: "error",
          error: "No se encontró el producto.",
        });
      } else {
        setProductData({
          data: data,
          status: "fullfilled",
          error: "",
        });
      }
    } catch (error) {
      setProductData({
        data: null,
        status: "error",
        error: error.message,
      });
    }
  }

  return (
    <div className="App">
      <Header />
      <main>
        <section>
          <SearchBox
            title={
              productData.status === "idle" ? "Indicá el id del producto" : null
            }
            handleSearch={handleSearch}
          />
          <Product productData={productData} />
        </section>
      </main>
    </div>
  );
}

export default App;
