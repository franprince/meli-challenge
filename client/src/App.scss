main {
  background-color: #e7e7e7;
  display: flex;
  min-height: calc(100vh - 56px);
  justify-content: center;
  section {
    margin: 30px;
    padding: 30px;
    max-width: 600px;
    width: 100%;
    background-color: #fff;
    height: fit-content;
    box-shadow: 1px 1px 2px 1px rgba(0, 0, 0, 0.1);
    @media (max-width: 500px) {
      max-width: 270px;
    }
  }
}

.nav-header {
  background-color: #fff159;
  border: 0;
  position: relative;
  height: 56px;
  box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.1);
  .nav-container {
    padding-top: 10px;
    margin: 0 auto;
    display: flex;
    max-width: 1200px;
    justify-content: space-between;
    align-items: center;
    .nav-logo {
      background-image: url("./images/logo.png");
      height: 34px;
      top: 11px;
      width: 134px;
      background-repeat: no-repeat;
      overflow: hidden;
      text-indent: -999px;
      left: 10px;
    }
    .nav-help {
      text-decoration: none;
      font-size: 14px;
      &:visited {
        color: #000;
      }
    }
    @media (max-width: 1023px) {
      .nav-logo {
        background-image: url("./images/logo_mobile.png");
      }
      margin: 0 10px;
    }
  }
}

.search-container {
  background-color: #fff;
  height: fit-content;
  label {
    font-weight: 600;
    padding-bottom: 20px;
    font-size: 20px;
  }
  input {
    border: 1px solid rgba(0, 0, 0, 0.2);
    width: calc(100% - 120px);
    padding: 10px;
    margin-right: 15px;
    &:focus {
      border: 1px solid rgba(0, 0, 0, 0.4);
      outline: 0;
    }
  }
  button {
    font-weight: 600;
    padding: 10px 15px;
    border-radius: 5px;
    border-color: transparent;
    background-color: rgba(65, 137, 230, 0.15);
    color: #3483fa;
    &:hover {
      cursor: pointer;
      background-color: rgba(65, 137, 230, 0.2);
      transition: box-shadow 0.25s ease-out, background-color 0.2s ease-out;
      box-shadow: 0 0 0 0 #fff;
    }
  }
  .search-form {
    display: flex;
    flex-direction: column;
  }
}
.product-box {
  display: flex;
  flex-direction: column;

  .product-id {
    font-size: 12px;
    color: rgba(0, 0, 0, 0.3);
    align-self: flex-end;
    margin: 20px 0 5px 0;
    width: 50%;
  }
  .product-name {
    flex: 50%;
    font-size: 20px;
    color: rgba(0, 0, 0, 0.8);
    font-weight: 600;
    padding-bottom: 8px;
    line-height: 1.18;
    word-break: break-word;
    align-self: flex-end;
    margin: 0;
    width: 50%;
  }
  .product-wrapper {
    display: flex;
  }
  .product-gallery-container {
    display: flex;
    flex: 50%;
    align-items: center;

    .product-image-container {
      display: flex;
      flex: 1;
      justify-content: center;
      .product-image {
        max-width: auto;
        max-height: 200px;
      }
    }

    .thumbnail-container {
      display: flex;
      flex-direction: column;
      justify-self: start;
      .product-thumbnail {
        max-height: 30px;
        border: 1px solid grey;
        padding: 3px;
        margin: 2px 0;
      }
    }
  }
  .product-data-container {
    flex: 50%;
    ul {
      padding: 0 20px;
      li {
        line-height: 30px;
        color: rgba(0, 0, 0, 0.6);
      }
    }
  }
  .toggle-more-info {
    align-self: start;
    background: none;
    border: none;
    color: #3483fa;
    padding: 30px 0 0 0;
    &:hover {
      cursor: pointer;
    }
  }
  @media (max-width: 500px) {
    .product-name {
      margin: 20px 0 5px 0;
      width: 100%;
    }
    .product-id {
      width: 100%;
    }
    .product-wrapper {
      flex-direction: column;
    }
    .product-gallery-container {
      flex-direction: column-reverse;
      .thumbnail-container {
        flex-direction: row;
        width: 100%;
        white-space: nowrap;
        overflow: auto;
        justify-content: center;
        .product-thumbnail {
          max-height: 50px;
          border: 1px solid grey;
          padding: 3px;
          margin: 2px 5px;
        }
      }
    }
  }
}
.product-more-details {
  transition: all 500ms;
  ul {
    padding: 0 20px;
    li {
      line-height: 30px;
      color: rgba(0, 0, 0, 0.6);
    }
  }
}
