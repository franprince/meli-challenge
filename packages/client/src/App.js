import "./App.scss";
import React, { useState } from "react";
import Product from "./components/Product";
import Header from "./components/Header";
import SearchBox from "./components/SearchBox";

function App() {
  const [productData, setProductData] = useState({
    data: null,
    status: "idle",
    error: "",
  });

  async function handleSearch(event, value) {
    event.preventDefault();
    try {
      if (!value) return;
      setProductData({ ...productData, status: "pending" });
      const res = await fetch(`http://localhost:3001/products/${value}`);
      const data = await res.json();
      if (res.status === 404) {
        setProductData({
          data: null,
          status: "error",
          error: "No se encontró el producto.",
        });
      } else if (res.status === 200) {
        setProductData({
          data: data,
          status: "fullfilled",
          error: "",
        });
      } else {
        setProductData({
          data: null,
          status: "error",
          error: res.status === 500 ? "Error del servidor" : "Hubo un error",
        });
      }
    } catch (error) {
      setProductData({
        data: null,
        status: "error",
        error: error.message,
      });
    }
  }

  return (
    <div className="App">
      <Header />
      <main>
        <section>
          <SearchBox
            title={
              productData.status === "idle" ? "Indicá el id del producto" : null
            }
            handleSearch={handleSearch}
          />
          <Product productData={productData} />
        </section>
      </main>
    </div>
  );
}

export default App;
