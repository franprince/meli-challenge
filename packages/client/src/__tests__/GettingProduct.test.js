import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

beforeEach(() => render(<App />));

test("it finds and displays a product", async () => {
  userEvent.type(
    screen.getByLabelText("Indicá el id del producto"),
    "MLA14186171"
  );
  userEvent.click(screen.getByRole("button", { name: /buscar/i }));
  expect(screen.getByText("Buscando...")).toBeInTheDocument();
  await waitFor(() =>
    expect(
      screen.getByText(/samsung galaxy s10 512 gb azul prisma 8 gb ram/i)
    ).toBeInTheDocument()
  );
  expect(
    screen.getByText(/código del catálogo: MLA14186171/i)
  ).toBeInTheDocument();
  expect(screen.getAllByRole("listitem")).toHaveLength(5);
  expect(screen.getAllByAltText(/miniatura/i)).toHaveLength(5);
});

test("it doesn't find a product and shows a message", async () => {
  userEvent.type(screen.getByLabelText("Indicá el id del producto"), "error");
  userEvent.click(screen.getByRole("button", { name: /buscar/i }));
  expect(screen.getByText(/buscando.../i)).toBeInTheDocument();
  await waitFor(() =>
    expect(screen.getByText(/no se encontró el producto./i)).toBeInTheDocument()
  );
});

test("interface doesn't change when you search with an empty input", async () => {
  userEvent.click(screen.getByRole("button", { name: /buscar/i }));
  const { container } = render(<App />);
  expect(container.firstChild).toMatchSnapshot();
});
