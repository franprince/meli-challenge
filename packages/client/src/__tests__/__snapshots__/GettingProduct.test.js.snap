// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interface doesn't change when you search with an empty input 1`] = `
<div
  class="App"
>
  <header
    class="nav-header"
  >
    <div
      class="nav-container"
    >
      <a
        class="nav-logo"
        href="/"
      >
        Volver al home
      </a>
      <a
        class="nav-help"
        href="https://www.mercadolibre.com.ar/ayuda"
      >
        Ayuda
      </a>
    </div>
  </header>
  <main>
    <section>
      <div
        class="search-container"
      >
        <form
          class="search-form"
        >
          <label
            for="searchById"
          >
            Indic√° el id del producto
          </label>
          <div
            class="input-container"
          >
            <div
              class="input-with-icons"
            >
              <input
                id="searchById"
                placeholder="EJ: MLA14186171"
                type="text"
                value=""
              />
              <button
                class="search-button"
              >
                Buscar
              </button>
              <button
                class="icon-button-search"
              >
                <svg
                  clip-rule="evenodd"
                  fill="gray"
                  fill-rule="evenodd"
                  height="24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15.853 16.56c-1.683 1.517-3.911 2.44-6.353 2.44-5.243 0-9.5-4.257-9.5-9.5s4.257-9.5 9.5-9.5 9.5 4.257 9.5 9.5c0 2.442-.923 4.67-2.44 6.353l7.44 7.44-.707.707-7.44-7.44zm-6.353-15.56c4.691 0 8.5 3.809 8.5 8.5s-3.809 8.5-8.5 8.5-8.5-3.809-8.5-8.5 3.809-8.5 8.5-8.5z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </form>
      </div>
    </section>
  </main>
</div>
`;
